spec.bs

```diff
@@ -1546,6 +1563,24 @@ To <dfn>generate and score bids</dfn> given an [=auction config=] |auctionConfig
 1. Wait until both |pendingBuyers| and |pendingAdditionalBids| are 0.
 1. If |leadingBidInfo|'s [=leading bid info/leading bid=] is null, return null.
 1. If |topLevelAuctionConfig| is null:
+  1. [=list/For each=] |generateBidResult| of |generateBidResults|:
+    1. If |generateBidResult|'s [=generate bid result/generated bid=] is |leadingBidInfo|'s

```

Comment:
This comment seems obsolete now, can you point me to the right place in the diff I should be looking at?

Reply:
https://github.com/WICG/turtledove/pull/977/files#diff-6f5a1d8263b0b0c42e2716ba5750e3652e359532647ac934c1c70086ae3ceddaR2291.
Basically it's about how to compare two structs (or, can we use |struct1| is |struct2| to compare two structs).

Reply:
I would say "if |struct1| is equal to |struct2|, for member-wise comparison" or something like that. I think you'd want to make the member-wise part more clear. To be extra formal, you would want to specify an equality algorithm that does the member-wise comparison explicitly. This is more important if the members are not primitives. That's kind of what https://html.spec.whatwg.org/multipage/browsers.html#same-origin does when comparing origins, which are structs.

But if the members are primitive, I could stomach settling with "member-wise comparison" terminology.

Reply:
I added an "id" field to [=generated bid=] instead since it seems easier.
It needs to add a lot if for equality algorithms, for example, [=generated bid=]'s equality algorithm needs another algorithm for [=interest group=]'s equality, etc,.

Reply:
When do multiple generated bids ever have the same ID then (so that they can be seen as "equal")? Do they get the same IDs if they are member-wise equivalent?

Reply:
no multiple generated bids will have the same ID. It shouldb unique.
After auction, I have a list of [=generated bids=] for all buyers, and a [=leading bid info=] (the auction winner's info) which has a field [=generated bid=] (the [=generated bid=] in the list that wins the auction). What I needed was a way to check which generated bid in the list is the one stored in [=leading bid info=], to determine which [=generated bid=]'s win report should be sent.
So the purpose was not to determine equality, but to check the [=leading bid info=]'s [=leading bid=] (a [=generated bid=]) is which one in a list of [=generated bids=].

Reply:
it's set [here](https://github.com/WICG/turtledove/pull/977/files#diff-6f5a1d8263b0b0c42e2716ba5750e3652e359532647ac934c1c70086ae3ceddaR1510) as the current size of a list of generated bids, and the list is growing when there's a new generated bid. I think there's no race condition (multiple component auctions running in parallel shouldn't have problem updating the list)?

Reply:
Done.

Reply:
@domfarolino Mind taking a final look at whether this is resolved? I think it was the last open comment that needs to be addressed.

Reply:
It's quite different in implementation. We have a class InterestGroupAuction representing an auction (in multi-SSP auction, each component auction will have its own InterestGroupAuction), which handles everything for its own auction. And each InterestGroupAuction has an array member bid_states, which is a list of [BidStates](https://crsrc.org/c/content/browser/interest_group/interest_group_auction.h;drc=5e4575c404a5571e5997be05352ab16fdf912361;l=175), one per buyer of this auction created when loading IGs. So during auction, there's no need to create |generatedBid| then append it to a |generatedBids|, but just need to update corresponding bid_state. After the whole auction finishes, it loops through InterestGroupAuctions (top level and component auctions) to collect debug reports.
In the spec, we currently don't have such an InterestGroupAuction class, so I created a global |generatedBids| that holds all auctions' |generatedBid|. But it would be best or unavoidable to follow what our code does, i.e., create an auction class to hold each auction's data separately. It will be required when we start adding post auction signals (e.g., winning-bid) which is auction-wise that each auction has their own such signals that need to be stored separately for each auction. I'll create a github issue and link to it in the spec.